File: src/main/java/sage/entity/Tweet.java
package sage.entity;

import java.util.*;
import javax.persistence.*;

import sage.domain.commons.IdCommons;
import sage.transfer.MidForwards;

@Entity
public class Tweet {
  private Long id;
  private String content;
  private User author;
  private Date time;
  private Long timeMillis;

  private Long originId = -1L;

  private String midForwardsJson = null;
  private Long blogId = -1L;
  private Set<Tag> tags = new HashSet<>();
  private Collection<Comment> comments = new ArrayList<>();

  private boolean deleted;

  Tweet() {}

  public Tweet(String content, User author, Date time, Set<Tag> tags) {
    this.content = content;
    this.author = author;
    this.time = time;
    this.timeMillis = time.getTime();
    this.tags.addAll(tags);
  }

  public Tweet(String content, User author, Date time, Tweet initialOrigin) {
    this(content, author, time, initialOrigin.getTags());
    setOriginId(initialOrigin.getId());
    if (initialOrigin.hasOrigin()) {
      throw new IllegalArgumentException(String.format(
          "tweet's origin should not be nested! initialOrigin[%s] and its origin[%s]",
          initialOrigin.getId(), initialOrigin.getOriginId()));
    }
  }

  public Tweet(String content, User author, Date time, Tweet origin, MidForwards midForwards) {
    this(content, author, time, origin);
    setMidForwardsJson(midForwards.toJson());
  }

  public Tweet(String content, User author, Date time, Blog sourceBlog) {
    this(content, author, time, sourceBlog.getTags());
    setBlogId(sourceBlog.getId());
  }

  @Id
  @GeneratedValue
  public Long getId() {
    return id;
  }
  public void setId(Long id) {
    this.id = id;
  }

  @Column(columnDefinition = "TEXT")
  @Lob
  public String getContent() {
    if (isDeleted()) return "";
    return content;
  }
  public void setContent(String content) {
    this.content = content;
  }

  @ManyToOne(optional = false)
  public User getAuthor() {
    return author;
  }
  public void setAuthor(User author) {
    this.author = author;
  }

  public Date getTime() {
    return time;
  }
  public void setTime(Date time) {
    this.time = time;
  }

  public Long getTimeMillis() {
    return timeMillis;
  }
  public void setTimeMillis(Long timeMillis) {
    this.timeMillis = timeMillis;
  }

  @Column(nullable = false)
  public Long getOriginId() {
    if (isDeleted()) return -1L;
    return originId;
  }
  public void setOriginId(Long originId) {
    this.originId = originId;
  }
  public boolean hasOrigin() {
    return originId >= 0;
  }

  public String getMidForwardsJson() {
    if (isDeleted()) return null;
    return midForwardsJson;
  }
  public void setMidForwardsJson(String midForwardsJson) {
    this.midForwardsJson = midForwardsJson;
  }

  @Column(nullable = false)
  public Long getBlogId() {
    return blogId;
  }
  public void setBlogId(Long blogId) {
    this.blogId = blogId;
  }

  @ManyToMany(fetch = FetchType.EAGER)
  public Set<Tag> getTags() {
    return tags;
  }
  public void setTags(Set<Tag> tags) {
    this.tags = tags;
  }

  @OneToMany(mappedBy = "source")
  public Collection<Comment> getComments() {
    return comments;
  }
  public void setComments(Collection<Comment> comments) {
    this.comments = comments;
  }

  public boolean isDeleted() {
    return deleted;
  }
  public void setDeleted(boolean deleted) {
    this.deleted = deleted;
  }

  public boolean hasBlog() {
    return blogId >= 0;
  }

  public MidForwards midForwards() {
    String json = getMidForwardsJson();
    return json == null ? null : MidForwards.fromJson(json);
  }

  @Override
  public String toString() {
    return author + ": " + content + tags;
  }

  @Override
  public int hashCode() {
    return IdCommons.hashCode(getId());
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    
    Tweet other = (Tweet) obj;
    return IdCommons.equal(getId(), other.getId());
  }
}

File: src/main/kotlin/sage/entity/Tweet.kt
package sage.entity

import org.slf4j.LoggerFactory
import sage.domain.commons.Edge
import sage.domain.commons.RichElement
import sage.domain.constraints.TweetConstraints
import sage.transfer.MidForward
import sage.transfer.MidForwards
import sage.web.context.Json
import java.util.*
import javax.persistence.*

@Entity
class Tweet : BaseModel {

  @Column(columnDefinition = "TEXT")
  @Lob
  val inputContent: String
    get() = if (deleted) "" else field

  @Column(columnDefinition = "TEXT")
  @Lob @Basic
  val content: String
    get() = if (deleted) "" else field

  @Column(columnDefinition = "TEXT")
  @Lob @Basic
  var richElementsJson: String?
    get() = if (deleted) null else field

  @Column(columnDefinition = "TEXT")
  @Lob @Basic
  var midForwardsJson: String? = null
    get() = if (deleted) null else field

  @ManyToOne(optional = false)
  val author: User

  @Column(nullable = false)
  var originId: Long = 0

  @Column(nullable = false)
  var blogId: Long = 0

  @ManyToMany(cascade = arrayOf(CascadeType.ALL))
  var tags: MutableSet<Tag> = HashSet()

  var deleted: Boolean = false

  constructor(inputContent: String, hyperContent: String, richElements: Collection<RichElement>, author: User, tags: Set<Tag>) {
    this.inputContent = inputContent
    this.content = hyperContent
    this.author = author
    this.richElementsJson = Json.json(richElements)
    this.tags = HashSet(tags)
  }

  constructor(inputContent: String, hyperContent: String, richElements: Collection<RichElement>, author: User, initialOrigin: Tweet)
      : this(inputContent, hyperContent, richElements, author, initialOrigin.tags) {
    originId = initialOrigin.id
    if (initialOrigin.hasOrigin()) {
      throw IllegalArgumentException(String.format(
          "tweet's origin should not be nested! initialOrigin[%s] and its origin[%s]",
          initialOrigin.id, initialOrigin.originId))
    }
  }

  constructor(inputContent: String, hyperContent: String, richElements: Collection<RichElement>, midForwards: MidForwards, author: User, origin: Tweet)
      : this(inputContent, hyperContent, richElements, author, origin) {
    midForwardsJson = midForwards.toJson()
  }

  constructor(inputContent: String, hyperContent: String, richElements: Collection<RichElement>, author: User, blog: Blog)
      : this(inputContent, hyperContent, richElements, author, blog.tags) {
    blogId = blog.id
  }

  fun hasOrigin() = originId > 0

  fun hasBlog() = blogId > 0

  fun richElements(): Collection<RichElement> =
      try {
        if (richElementsJson != null && richElementsJson != "[]") RichElement.fromJsonToList(richElementsJson!!)
        else emptyList()
      } catch (e: Exception) {
        log.error("richElements cannot be deserialized from JSON", e)
        emptyList<RichElement>()
      }

  fun midForwards(): MidForwards? =
    try {
      midForwardsJson?.run { MidForwards.fromJson(this) }
    } catch (e: Exception) {
      log.error("midForwards cannot be deserialized from JSON", e)
      MidForwards().apply { xs.add(MidForward(0, 0, "", "")) }
    }

  fun stat() = TweetStat.byId(id)

  fun validate() {
    TweetConstraints.check(this)
  }

  companion object : Find<Long, Tweet>() {
    private val log = LoggerFactory.getLogger(Tweet::class.java)

    fun get(id: Long): Tweet = getNonNull(Tweet::class, id)

    fun recent(size: Int): List<Tweet> = default().orderBy("id desc").setMaxRows(size).findList()

    fun byTags(tags: Collection<Tag>, edge: Edge): List<Tweet> {
      if (tags.isEmpty()) {
        return LinkedList()
      }
      return ranged(edge).`in`("tags.id", Tag.getQueryTags(tags).map(Tag::id)).findList()
    }

    fun byAuthor(authorId: Long, edge: Edge = Edge.none()): MutableList<Tweet> = ranged(edge).eq("author", User.ref(authorId)).findList()

    fun byAuthorAndTags(authorId: Long, tags: Collection<Tag>, edge: Edge = Edge.none()): List<Tweet> {
      if (tags.isEmpty()) {
        return LinkedList()
      }
      if (hasRoot(tags)) {
        return byAuthor(authorId, edge)
      }
      return ranged(edge).eq("author", User.ref(authorId)).`in`("tags.id", Tag.getQueryTags(tags).map(Tag::id)).findList()
    }

    fun forwards(originId: Long): List<Tweet> = default().eq("originId", originId).findList()

    fun getOrigin(tweet: Tweet) = if (tweet.hasOrigin()) byId(tweet.originId) else null

    private fun ranged(edge: Edge) = default().apply {
      when (edge.type) {
        Edge.EdgeType.NONE -> {}
        Edge.EdgeType.BEFORE -> lt("id", edge.edgeId)
        Edge.EdgeType.AFTER -> gt("id", edge.edgeId)
        else -> throw UnsupportedOperationException()
      }
      orderBy("id desc").maxRows = edge.limitCount
    }

    private fun default() = where().eq("deleted", false)

    private fun hasRoot(tags: Collection<Tag>) = tags.find { it.id == Tag.ROOT_ID } != null
  }
}